### PR TITLE
[NativeAOT] Adding CET support

### LINKED ISSUE #101942: [NativeAOT] Enable CET support
CET is one of the low-level [security mitigations](https://github.com/dotnet/designs/blob/main/accepted/2021/runtime-security-mitigations.md).

It is supported on CoreCLR, but not on native AOT. We need to port the support from CoreCLR to native AOT (and test it that it works).

(Context https://github.com/dotnet/runtime/pull/101891#issuecomment-2096788578)

### PR DESCRIPTION
**Pull request by VSadov**
Fixes: https://github.com/dotnet/runtime/issues/101942

- Added support for `STATUS_RETURN_ADDRESS_HIJACK_ATTEMPT`
- Set /CETCOMPAT in executables by default, unless `/CETCOMPAT:NO`
- Opt into EHCONT (if CFG is enabled)
- Reconcile shadow stack with SP changes in `RhpCallCatchFunclet`

### CODE REVIEW

File: src/coreclr/vm/excep.cpp [hunk 1]
```diff
@@ -6767,10 +6767,12 @@ VEH_ACTION WINAPI CLRVectoredExceptionHandler(PEXCEPTION_POINTERS pExceptionInfo
 
     if (pExceptionInfo->ExceptionRecord->ExceptionCode == STATUS_RETURN_ADDRESS_HIJACK_ATTEMPT)
     {
-        if (pThread == NULL || !pThread->PreemptiveGCDisabled())
+        if (pThread == NULL ||
+            !pThread->PreemptiveGCDisabled() ||
+            !pThread->HasThreadState(Thread::TS_Hijacked))
```
Comment (jkotas) [inline src/coreclr/vm/excep.cpp:6772]: What is the situation that the `TS_Hijacked` check is trying to handle?

Comment (VSadov) [reply src/coreclr/vm/excep.cpp:6772]: If we did not hijack the thread, we cannot continue here. We can't unhijack it. 
If we see such state, we can either crash or just let this be handled by whoever did the hijack. 

There is a theoretical possibility that someone else hijacked our thread, even if we were running in coop mode. Managed runtimes (dotnet or whatever) might not do that, but there is still a possibility. Maybe something like debugger or profiler could do this. 

This style of hijacking does not allow to tell who really hijacked a thread. On the other hand if a thread was hijacked twice, by different entities, I think things could work.

Comment (jkotas) [reply src/coreclr/vm/excep.cpp:6772]: Two entities trying to hijack the same return address would be a factory for race conditions. If we detect that somebody else tried to hijack our return address, I think the best action is immediate fail fast.

Comment (VSadov) [reply src/coreclr/vm/excep.cpp:6772]: fail fast would work too. 

One thing is certain - if the hijacked thread (or more precisely - the hijacked location) was not hijacked by us, we cannot continue on this path.

File: src/coreclr/nativeaot/Runtime/StackFrameIterator.cpp [hunk 2]
```diff
@@ -616,6 +617,29 @@ void StackFrameIterator::InternalInit(Thread * pThreadToWalk, NATIVE_CONTEXT* pC
...
 #undef PTR_TO_REG
+
+    // This function guarantees that the final initialized context will refer to a managed
+    // frame.  In the rare case where the PC does not refer to managed code (and refers to an
+    // assembly thunk instead), unwind through the thunk sequence to find the nearest managed
+    // frame.
+    // NOTE: When thunks are present, the thunk sequence may report a conservative GC reporting
+    // lower bound that must be applied when processing the managed frame.
+
+    ReturnAddressCategory category = CategorizeUnadjustedReturnAddress(m_ControlPC);
```
Comment (janvorli) [inline src/coreclr/nativeaot/Runtime/StackFrameIterator.cpp:628]: I am not sure how this is related to the CET stuff

Comment (VSadov) [reply src/coreclr/nativeaot/Runtime/StackFrameIterator.cpp:628]: InternalInit is supposed to leave the iterator in managed code. Until this change the InternalInit that takes `NATIVE_CONTEXT` was always starting from a location that is already in managed code.  (unlike the one that takes `PInvokeTransitionFrame`)

Now, when we receive hijack exception, we pop to the caller and perform suspension in the caller's context. 
Typically managed method that is not a reverse PInvoke must only be called by another managed method, so that would require no changes here. There are rare cases though in NativeAOT when ordinary managed methods are called from asm thunks. So it becomes possible for this code to see a location in a thunk. That is handled in the same way as in the other InternalInit method. (i.e. via `UnwindNonEHThunkSequence`)

Perhaps I should unify this code into a helper to not duplicate it in both InternalInit methods.

Comment (VSadov) [reply src/coreclr/nativeaot/Runtime/StackFrameIterator.cpp:620]: I've unified the adjustment for asm thunks

**Comment by VSadov** [comment]
Turns out my CET-compatible machine already enables shadow stacks for executables that are CET-compatible. (i.e. msedge.exe, conhost.exe, ...).  

If we pass `/CETCOMPAT` to linker, NativeAOT executables are also opted in. And it looks like passing tests, while running in `Compatible modules only` mode. (same as msedge.exe).

The next step is to make sure we are not relying for some compat/fixup behavior in exception handling, and whether we need `/EHCONT`

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Windows.targets [hunk 3]
```diff
@@ -99,6 +99,10 @@ The .NET Foundation licenses this file to you under the MIT license.
       <LinkerArg Condition="'$(OutputType)' == 'WinExe' or '$(OutputType)' == 'Exe'" Include="/STACK:$(IlcDefaultStackSize)" />
       <!-- Do not warn if someone declares UnmanagedCallersOnly with an entrypoint of 'DllGetClassObject' and similar -->
       <LinkerArg Include="/IGNORE:4104" />
+      <!-- Opt into CETCOMPAT by default. -->
+      <LinkerArg Condition="'$(ShadowStacksCompatible)' != 'false'" Include="/CETCOMPAT" />
```
Comment (jkotas) [inline src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Windows.targets:103]: We should call this property `CETCompat`. It is what it is called in .vcxproj files.

Comment (VSadov) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Windows.targets:103]: Also, I think we may want make it x64-only, just in case it starts doing something for x86. 
According to docs it is accepted for x86 and currently does nothing.

Comment (jkotas) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Windows.targets:103]: > I think we may want make it x64-only, 

Sounds good to me.

Comment (jkotas) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Windows.targets:103]: (I will start doing something for arm64 eventually.)

Comment (VSadov) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Windows.targets:103]: Also `/CETCOMPAT` seems not accepted on arm64

**Comment by VSadov** [comment]
`/guard:ehcont` is a bit tricky. Once that is set the things like following start failing RIP validation in `NtContinue`.

```c
    if (translateToManagedException)
    {
        pExPtrs->ContextRecord->SetIp(PCODEToPINSTR((PCODE)&RhpThrowHwEx));
        pExPtrs->ContextRecord->SetArg0Reg(faultCode);
        pExPtrs->ContextRecord->SetArg1Reg(faultingIP);

        return EXCEPTION_CONTINUE_EXECUTION;
    }
```

**Comment by VSadov** [comment]
The other interesting part is that `/guard:ehcont` requires CFG which we do not enable by default, as there are reasons.
One possibility is to enable `/guard:ehcont` only when CGF is already enabled. 

For now I will assume that we will be enabling `/guard:ehcont` one way or another. We need to make the scenario work before settling under which conditions it is enabled,

File: src/coreclr/vm/gcinfodecoder.cpp [hunk 4]
```diff
@@ -786,7 +786,7 @@ bool GcInfoDecoder::EnumerateLiveSlots(
 
                 UINT32 normStart = lastNormStop + normStartDelta;
                 UINT32 normStop = normStart + normStopDelta;
-                if(normBreakOffset >= normStart && normBreakOffset < normStop)
+                if(normBreakOffset >= normStart && normBreakOffset <= normStop)
```
Comment (VSadov) [inline src/coreclr/vm/gcinfodecoder.cpp:789]: This is an ugly hack to workaround a rare case where a call to a hijacked method is immediately followed by a NoGC region thus the caller IP is outside of interruptible code. 
It is always safe to pop to the caller of a regular managed method and start stackwalk at the call site, but when we do that as a part of hijack and the above condition present, it triggers asserts if caller is also fully-interruptible.

A proper fix for this condition is in PR: https://github.com/dotnet/runtime/pull/103540

**Comment by VSadov** [comment]
I think the technical part of this PR is complete - we can enable CET for NativeAOT. What remains it to settle on enablement policy and how to test this.

What I can suggest for enablement:
- The basic opt-in to CET could happen unconditionally. This is the part where we mark binary as CET compatible and if OS enables CET, we use different APIs for hijacking.
This is the state that mitigates typical ROP attacks, but things like exception handling could work as before and rely on OS fixups (we would not be relying on that).
In this mode the actual binary is roughly the same and the perf should be roughly the same.

- The next level of opt-in is opting into EHCONT 
The inconvenient part here is that this will also require opting into CFG, which has known costs in terms of size and throughput.
I can suggest to opt into EHCONT if user has already opted into CFG.

Basically:
- If user explicitly opts out of CET (as in `/CETCOMPAT:NO`), then nothing is enabled. Binary is not marked for CET
- By default the binary is marked as CET compatible, but not opted into EHCONT
- If CFG is enabled, we also opt into EHCONT.

Any other ideas?
@jkotas  @janvorli

**Comment by VSadov** [comment]
For the testing part - it looks like we have a bunch of tests that are incompatible with CFG. 
A PR without any of these changes, but with forcing CFG (https://github.com/dotnet/runtime/pull/103215) demonstrates that Libraries tests mostly run, but smoke tests mostly fail. 

We need to do something about that - either make tests compatible with CFG (where possible) or disable some tests/scenarios if can't made compatible.  
Perhaps have a leg in `runtime-nativeaot-outerloop` that enables CFG?

@MichalStrehovsky

**Comment by jkotas** [comment]
> I am basically optimizing for minimum number of knobs and assume that CET without EHCONT could be a good default.

Sounds good to me.

**Comment by jkotas** [comment]
> For the testing part - it looks like we have a bunch of tests that are incompatible with CFG.

At least some of these failures look like product bugs to me. We want to fix them - CFG is a supported config for naot. Have you looked into where the problems are?

>Perhaps have a leg in runtime-nativeaot-outerloop that enables CFG?

I agree that we can use more outer loops tests with CFG enabled given these results.

**Comment by VSadov** [comment]
>> For the testing part - it looks like we have a bunch of tests that are incompatible with CFG.

> At least some of these failures look like product bugs to me. We want to fix them - CFG is a supported config for naot. Have you looked into where the problems are?

Not very deeply. I was mostly concerned that CET does not introduce anything new.

The first few failures that I looked at were explicitly introspecting the native code bytes (to be sure that constants are emitted as constants and the like) and those were obviously sensitive to CFG.  
Also by looking at Libraries tests passing with forced CFG, I think we are mostly compatible, so failures looked as more likely to be test issues. 
There could be bugs too.

**Comment by agocke** [comment]
> I am basically optimizing for minimum number of knobs and assume that CET without EHCONT could be a good default.

Agreed

**Comment by MichalStrehovsky** [comment]
> Perhaps have a leg in `runtime-nativeaot-outerloop` that enables CFG?

That sounds good to me. CFG is getting tested with CoreCLR in RyuJIT legs since it's mostly a codegen feature. We have one smoke test to validate we're passing the right flag to JIT to activate it and check we mark all method bodies as entrypoints, basically. Could be better.

We'll definitely get more in terms of bang for the buck if it's combined with CET testing.

**Comment by MichalStrehovsky** [comment]
> > At least some of these failures look like product bugs to me. We want to fix them - CFG is a supported config for naot. Have you looked into where the problems are?
> 
> Not very deeply. I was mostly concerned that CET does not introduce anything new.

I looked at the failure in the Exceptions smoke test. There is _some_ bug it found, but it's not clear to me where the bug is. Filed it as #103798.

**Comment by VSadov** [comment]
Just for comparison enabling CFG changes the size of Release `System.Collections.Concurrent.Tests.exe`  

from: 20,788,224 bytes
to: 25,157,120 bytes

That is 21% increase

**Comment by VSadov** [comment]
I think this is ready for review/merge.

**Comment by VSadov** [comment]
The only remaining follow up is to: 
- create a CFG-enabled test leg, which will also test CET, if run on CET-capable machines. 
- fix CFG-related bugs (i.e. https://github.com/dotnet/runtime/issues/103798 or whatever else is found)

I think that should be done outside of this PR.

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Windows.targets [hunk 5]
```diff
@@ -99,6 +99,12 @@ The .NET Foundation licenses this file to you under the MIT license.
       <LinkerArg Condition="'$(OutputType)' == 'WinExe' or '$(OutputType)' == 'Exe'" Include="/STACK:$(IlcDefaultStackSize)" />
       <!-- Do not warn if someone declares UnmanagedCallersOnly with an entrypoint of 'DllGetClassObject' and similar -->
       <LinkerArg Include="/IGNORE:4104" />
+      <!-- Opt into CETCOMPAT by default. -->
+      <LinkerArg Condition="'$(CETCompat)' != 'false' and '$(_targetArchitecture)' == 'x64'" Include="/CETCOMPAT" />
+      <!-- Allow user to opt out. -->
+      <LinkerArg Condition="'$(CETCompat)' == 'false' and '$(_targetArchitecture)' == 'x64'" Include="/CETCOMPAT:NO" />
```
Comment (jkotas) [inline src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Windows.targets:105]: Is this future proofing in case CETCOMPAT is enabled by default by linker? (I do not think it is needed, but it does not hurt.)

File: src/coreclr/nativeaot/Runtime/EHHelpers.cpp [hunk 6]
```diff
@@ -536,6 +536,57 @@ int32_t __stdcall RhpVectoredExceptionHandler(PEXCEPTION_POINTERS pExPtrs)
...
+        Thread * pThread = ThreadStore::GetCurrentThreadIfAvailable();
+        if (pThread == NULL || !pThread->IsCurrentThreadInCooperativeMode())
+        {
+            // if we are not in coop mode, this cannot be our hijack
+            // Perhaps some other runtime is responsible.
+            return EXCEPTION_CONTINUE_SEARCH;
+        }
+
+        // Sanity check. The thread should be hijacked by us.
+        _ASSERTE_ALL_BUILDS(pThread->IsHijacked());
```
Comment (jkotas) [inline src/coreclr/nativeaot/Runtime/EHHelpers.cpp:557]: `_ASSERTE_ALL_BUILDS` is a regular debug-only assert in NAOT, so this does not do what it seems to be doing.

File: src/coreclr/nativeaot/Runtime/inc/CommonTypes.h [hunk 7]
```diff
@@ -61,4 +61,12 @@ typedef struct _GUID {
 } GUID;
 #endif // FEATURE_EVENT_TRACE && !_INC_WINDOWS
 
+// Hijack funcs are not called, they are "returned to". And when done, they return to the actual caller.
+// Thus they cannot have any parameters or return anything.
+#if defined(HOST_X86) && !defined(HOST_UNIX)
+typedef void __fastcall HijackFunc();
+#else
+typedef void HijackFunc();
+#endif
```
Comment (jkotas) [inline src/coreclr/nativeaot/Runtime/inc/CommonTypes.h:70]:
```suggestion
typedef void HijackFunc();
```
We should not need __fastcall on the hijack func. Can we update the asm mangled name instead to avoid the ifdef?

File: src/coreclr/nativeaot/Runtime/windows/PalRedhawkMinWin.cpp [hunk 8]
```diff
@@ -327,6 +332,20 @@ REDHAWK_PALEXPORT HANDLE REDHAWK_PALAPI PalCreateEventW(_In_opt_ LPSECURITY_ATTR
...
+REDHAWK_PALEXPORT UInt32_BOOL REDHAWK_PALAPI PalAreShadowStacksEnabled()
+{
+#if defined(TARGET_AMD64)
+        // The SSP is null when CET shadow stacks are not enabled. On processors that don't support shadow stacks, this is a
+        // no-op and the intrinsic returns 0. CET shadow stacks are enabled or disabled for all threads, so the result is the
+        // same from any thread.
+        return _rdsspq() != 0;
+#else
+        // When implementing AreShadowStacksEnabled() on other architectures, review all the places where this is used.
+        return false;
```
Comment (jkotas) [inline src/coreclr/nativeaot/Runtime/windows/PalRedhawkMinWin.cpp:344]:
```suggestion
    // The SSP is null when CET shadow stacks are not enabled. On processors that don't support shadow stacks, this is a
    // no-op and the intrinsic returns 0. CET shadow stacks are enabled or disabled for all threads, so the result is the
    // same from any thread.
    return _rdsspq() != 0;
#else
    // When implementing AreShadowStacksEnabled() on other architectures, review all the places where this is used.
    return false;
```
Nit: Indentiation?

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Windows.targets (hunk 5 above)
Comment (VSadov) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Windows.targets:105]: Just wanted to be explicit about this. But, yes, will also help if something defaults everything into "/CETCOMPAT"

File: src/coreclr/nativeaot/Runtime/EHHelpers.cpp [hunk 9]
```diff
@@ -536,6 +536,57 @@ int32_t __stdcall RhpVectoredExceptionHandler(PEXCEPTION_POINTERS pExPtrs)
...
+            _ASSERTE(*(size_t *)interruptedContext->GetSp() == (uintptr_t)pThread->GetHijackedReturnAddress());
+
+            // When the CET is enabled, the interruption happens on the ret instruction in the calee.
+            // We need to "pop" rsp to the caller, as if the ret has consumed it.
+            interruptedContext->SetSp(interruptedContext->GetSp() + 8);
+        }
+
+        // Change the IP to be at the original return site, as if we have returned to the caller.
+        // That IP is an interruptible safe point, so we can suspend right there.
+        uintptr_t origIp = interruptedContext->GetIp();
```
Comment (VSadov) [inline src/coreclr/nativeaot/Runtime/EHHelpers.cpp:573]: @jakobbotsch this style of hijacking can hit some asserts if the caller is missing safepoint for the call site. (i.e. MinOpt). The code will still work, but asserts will need to be adjusted.

Re: https://github.com/dotnet/runtime/pull/103950

Comment (jkotas) [reply src/coreclr/nativeaot/Runtime/EHHelpers.cpp:573]: Does this work the same way in regular CoreCLR?

Comment (VSadov) [reply src/coreclr/nativeaot/Runtime/EHHelpers.cpp:573]: It could. I think all hijacks could work this way and it will be simpler. 

I.E no need for special frames, interrogating the containing method info for kinds of GC returns, capturing that somewhere,…

Comment (VSadov) [reply src/coreclr/nativeaot/Runtime/EHHelpers.cpp:573]: That is assuming we bumped r2r version

Comment (jkotas) [reply src/coreclr/nativeaot/Runtime/EHHelpers.cpp:573]: > It could.

It is important that it does. Our native AOT testing strategy depends on minimizing differences between regular CoreCLR and native AOT.

Comment (VSadov) [reply src/coreclr/nativeaot/Runtime/EHHelpers.cpp:573]: Looks like r2r version was incremented. 👍

File: src/coreclr/nativeaot/Runtime/i386/GcProbe.asm [hunk 10]
```diff
@@ -261,16 +261,16 @@ WaitForGC:
...
+_RhpGcStressHijack@0  proc public
 
         HijackFixupProlog
         or          ecx, DEFAULT_PROBE_SAVE_FLAGS + PTFF_SAVE_RAX
         jmp         RhpGcStressProbe
 
-FASTCALL_ENDFUNC
+_RhpGcStressHijack@0  endp
 
 FASTCALL_FUNC RhpHijackForGcStress, 0
```
Comment (jkotas) [inline src/coreclr/nativeaot/Runtime/i386/GcProbe.asm:275]:
```suggestion
_RhpHijackForGcStress@0  proc public
```
This is currently unused, but I think it should match

Comment (VSadov) [reply src/coreclr/nativeaot/Runtime/i386/GcProbe.asm:275]: I was not sure if this would be called from C or asm. 

I hope to enable GC stress for NativeAOT in some form one day. 
Perhaps not in the form it was, as a lot of things changed with switch to RyuJIT and CoreCLR-like suspension plan. However, some of this code may still be useful at some point, so I try to keep it around, even if unused.

File: src/coreclr/nativeaot/Runtime/EHHelpers.cpp (hunk 9 above)
Comment (VSadov) [reply src/coreclr/nativeaot/Runtime/EHHelpers.cpp:577]: Since CET compat will be the default, this code path will be exercised on everything win10 and newer, so this will be reasonably tested. 

I can move CoreCLR on the same plan, so we have more confidence. I was planning to do that anyways.

Comment (VSadov) [reply src/coreclr/nativeaot/Runtime/EHHelpers.cpp:577]: Here a PR that uses this pattern in CoreCLR CET hijacks and in GC stress. 
https://github.com/dotnet/runtime/pull/104198

File: src/coreclr/nativeaot/Runtime/windows/PalRedhawkMinWin.cpp [hunk 11]
```diff
@@ -629,6 +651,58 @@ REDHAWK_PALEXPORT UInt32_BOOL REDHAWK_PALAPI PalRegisterHijackCallback(_In_ PalH
...
+    else
+#endif // HOST_AMD64
+    {
+        g_pfnQueueUserAPC2Proc = (QueueUserAPC2Proc)GetProcAddress(hKernel32, "QueueUserAPC2");
+    }
+
+    if (PalAreShadowStacksEnabled())
+    {
+        // When shadow stacks are enabled, support for special user-mode APCs is required
+        _ASSERTE_ALL_BUILDS(g_pfnQueueUserAPC2Proc != NULL);
```
Comment (jkotas) [inline src/coreclr/nativeaot/Runtime/windows/PalRedhawkMinWin.cpp:680]:
```suggestion
        _ASSERTE(g_pfnQueueUserAPC2Proc != NULL);
```

File: src/coreclr/nativeaot/Runtime/windows/PalRedhawkMinWin.cpp [hunk 12]
```diff
@@ -629,6 +651,58 @@ REDHAWK_PALEXPORT UInt32_BOOL REDHAWK_PALAPI PalRegisterHijackCallback(_In_ PalH
...
+            void* rtlGetReturnAddressHijackTarget = GetProcAddress(hModNtdll, "RtlGetReturnAddressHijackTarget");
+            if (rtlGetReturnAddressHijackTarget != NULL)
+            {
+                g_returnAddressHijackTarget = ((PFN_RtlGetReturnAddressHijackTarget)rtlGetReturnAddressHijackTarget)();
+            }
+        }
+
+        if (g_returnAddressHijackTarget == NULL)
+        {
+            _ASSERTE_ALL_BUILDS(!"RtlGetReturnAddressHijackTarget must provide a target when shadow stacks are enabled");
```
Comment (jkotas) [inline src/coreclr/nativeaot/Runtime/windows/PalRedhawkMinWin.cpp:696]:
```suggestion
            _ASSERTE(!"RtlGetReturnAddressHijackTarget must provide a target when shadow stacks are enabled");
```

File: src/coreclr/nativeaot/Runtime/windows/PalRedhawkMinWin.cpp [hunk 13]
```diff
@@ -629,6 +651,58 @@ REDHAWK_PALEXPORT UInt32_BOOL REDHAWK_PALAPI PalRegisterHijackCallback(_In_ PalH
...
+        if (hModNtdll != NULL)
+        {
+            typedef void* (*PFN_RtlGetReturnAddressHijackTarget)(void);
+
+            void* rtlGetReturnAddressHijackTarget = GetProcAddress(hModNtdll, "RtlGetReturnAddressHijackTarget");
+            if (rtlGetReturnAddressHijackTarget != NULL)
+            {
+                g_returnAddressHijackTarget = ((PFN_RtlGetReturnAddressHijackTarget)rtlGetReturnAddressHijackTarget)();
+            }
+        }
```
Comment (jkotas) [inline src/coreclr/nativeaot/Runtime/windows/PalRedhawkMinWin.cpp:692]:
```suggestion
        typedef void* (*PFN_RtlGetReturnAddressHijackTarget)(void);

        void* rtlGetReturnAddressHijackTarget = GetProcAddress(hModNtdll, "RtlGetReturnAddressHijackTarget");
        if (rtlGetReturnAddressHijackTarget != NULL)
        {
            g_returnAddressHijackTarget = ((PFN_RtlGetReturnAddressHijackTarget)rtlGetReturnAddressHijackTarget)();
        }
```
Nit: It is unnecessary to null check this. ntdll is always present; and if it is not - GetProcAddress handles NULL module gracefully.

**Review by jkotas — APPROVED** [review]
LGTM. Thank you!

**Comment by VSadov** [comment]
Thanks!!!
